driver/pelican/main.go

```diff
@@ -49,88 +54,102 @@ func main() {
-		}
-		fmt.Printf("%+v", state)
-		if state.Mode != nil {
-			m := float64(*state.Mode)
-			params.Mode = &m
-		}
-
-		if state.Override != nil && *state.Override {
-			f := float64(1)
-			params.Override = &f
-		} else {
-			f := float64(0)
-			params.Override = &f
-		}
-
-		if state.Fan != nil && *state.Fan {
-			f := float64(1)
-			params.Fan = &f
-		} else {
-			f := float64(0)
-			params.Fan = &f
-		}
-
-		if err := pelican.ModifyState(&params); err != nil {
-			fmt.Println(err)
-		} else {
-			fmt.Printf("Set Pelican state to: %+v\n", params)
-		}
-	})
-
-	for {
-		status, err := pelican.GetStatus()
-		if err != nil {
-			fmt.Printf("Failed to retrieve Pelican status: %v\n", err)
-			os.Exit(1)
-		}
-
-		po, err := bw2.CreateMsgPackPayloadObject(bw2.FromDotForm(TSTAT_PO_DF), status)
-		if err != nil {
-			fmt.Printf("Failed to create msgpack PO: %v", err)
-			os.Exit(1)
-		}
-		tstatIface.PublishSignal("info", po)
-		time.Sleep(pollInt)
+	tstatIfaces := make([]*bw2.Interface, len(pelicans))
+	for i, pelican := range pelicans {
+		tstatIfaces[i] = service.RegisterInterface(pelican.name, "i.xbos.thermostat")
+
+		tstatIfaces[i].SubscribeSlot("setpoints", func(msg *bw2.SimpleMessage) {
+			po := msg.GetOnePODF(TSTAT_PO_DF)
+			if po == nil {
+				fmt.Println("Received message on setpoints slot without required PO. Droping.")
+				return
+			}
+
+			var setpoints setpointsMsg
+			if err := po.(bw2.MsgPackPayloadObject).ValueInto(&setpoints); err != nil {
+				fmt.Println("Received malformed PO on setpoints slot. Dropping.", err)
+				return
+			}
+
+			if err := pelican.ModifySetpoints(&setpoints); err != nil {
+				fmt.Println(err)
+			} else {
+				fmt.Printf("Set heating setpoint to %v and cooling setpoint to %v\n",
+					setpoints.HeatingSetpoint, setpoints.CoolingSetpoint)
+			}
+		})
+
+		tstatIfaces[i].SubscribeSlot("state", func(msg *bw2.SimpleMessage) {
+			po := msg.GetOnePODF(TSTAT_PO_DF)
+			if po == nil {
+				fmt.Println("Received message on state slot without required PO. Dropping.")
+				return
+			}
+
+			var state stateMsg
+			if err := po.(bw2.MsgPackPayloadObject).ValueInto(&state); err != nil {
+				fmt.Println("Received malformed PO on state slot. Dropping.", err)
+				return
+			}
+
+			params := pelicanStateParams{
+				HeatingSetpoint: state.HeatingSetpoint,
+				CoolingSetpoint: state.CoolingSetpoint,
+			}
+			fmt.Printf("%+v", state)
+			if state.Mode != nil {
+				m := float64(*state.Mode)
+				params.Mode = &m
+			}
+
+			if state.Override != nil && *state.Override {
+				f := float64(1)
+				params.Override = &f
+			} else {
+				f := float64(0)
+				params.Override = &f
+			}
+
+			if state.Fan != nil && *state.Fan {
+				f := float64(1)
+				params.Fan = &f
+			} else {
+				f := float64(0)
+				params.Fan = &f
+			}
+
+			if err := pelican.ModifyState(&params); err != nil {
+				fmt.Println(err)
+			} else {
+				fmt.Printf("Set Pelican state to: %+v\n", params)
+			}
+		})
 	}
+
```

Comment:
I'm pretty sure the way this is implemented, the driver will continue to run until *all* Pelicans are no longer reachable (which can happen because they are wireless). I think it would be more robust if the driver restarted as soon as *any* of the Pelicans are more reachable. Maybe something like

```go
done := make(chan bool)
for i, pelican := range pelicans {
    currentPelican := pelican
    currentIface := tstatIfaces[i]
    go func() {
        for {
            status, err := currentPelican.GetStatus()
            if err != nil {
                fmt.Printf("Failed to retrieve Pelican status: %v\n", err)
                done <- true
            }

            po, err := bw2.CreateMsgPackPayloadObject(bw2.FromDotForm(TSTAT_PO_DF), status)
            if err != nil {
                fmt.Printf("Failed to create msgpack PO: %v", err)
                done <- true
            }
            currentIface.PublishSignal("info", po)
            time.Sleep(pollInt)
        }
    }()
}
<-done
```